src/Infiltrator.jl

```diff
@@ -159,57 +82,102 @@ const TEST_TERMINAL_REF = Ref{Any}(nothing)
 const TEST_REPL_REF = Ref{Any}(nothing)
 const TEST_NOSTACK = Ref{Any}(false)
 
-const EXITING = Ref{Bool}(false)
-const DISABLED = Set()
+mutable struct ScratchPad
+  store::Module
+  exiting::Bool
+  disabled::Set
+end
+function Base.show(io::IO, s::ScratchPad)
+  n = length(get_scratch_pad_names(s))
+  print(io, "Infiltrator scratch pad with $(n) name$(n == 1 ? "" : "s")")
+end
+function Base.getproperty(sp::ScratchPad, s::Symbol)
+  s === :store && return getfield(sp, :store)
+  s === :exiting && return getfield(sp, :exiting)
+  s === :disabled && return getfield(sp, :disabled)
+
+  if isdefined(sp.store, s)
+    getproperty(sp.store, s)
+  else
+    throw(UndefVarError(s))
+  end
+end
+Base.propertynames(s::ScratchPad) = keys(get_scratch_pad_names(s))
 
-const SCRATCH_PAD = Ref{Module}()
+"""
+    store
 
+Global scratch pad for storing values while `@infiltrate`ing or `@exfiltrate`ing.
 """
-    clear_disabled()
+const store = ScratchPad(Module(), false, Set())
+
+"""
+    @with ex
+
+Evaluates the expression `ex` in the context of the global scratch pad.
+
+Mainly intended for interactive use, as changes to the scratch pads state will not
+propagate into the returned expression.
+"""
+macro with(ex)
+  with_scratch_pad(ex)
+end
+
+function with_scratch_pad(ex)
+  ns = get_scratch_pad_names(store)
+  return Expr(:let,
+    Expr(:block, map(x->Expr(:(=), x...), [(k, maybe_quote(v)) for (k, v) in ns])...),
+    Expr(:block, ex)
+  )
+end
+
+"""
+    clear_disabled(s = store)
 
 Clear all disabled infiltration points.
 """
-function clear_disabled()
-  empty!(DISABLED)
+function clear_disabled(s = store)
+  empty!(s.disabled)
   return nothing
 end
 
 """
-    end_session()
+    end_session(s = store)
 
 End this infiltration session (reverts the effect of `@exit` in the `debug>` REPL).
 
 Only needs to be manually called on Julia versions prior to 1.5.
 """
-function end_session()
-  EXITING[] = false
+function end_session(s::ScratchPad = store)
+  s.exiting = false
   return nothing
 end
 
 """
-    clear_scratch_pad()
+    clear_store(s::ScratchPad = Infiltrator.store)
 
-Reset the scratch pad module used for global symbols.
+Reset the scratch pad used for global symbols.
 """
-function clear_scratch_pad()
-  SCRATCH_PAD[] = Module()
-  nothing
-end
+clear_store(s::ScratchPad = store) = set_store(s, Module())
 
 """
-    get_scratch_pad()
+    set_store(s::ScratchPad = Infiltrator.store, m::Module)
```

Comment:
Add an exclamation mark? Since it's modifying `s`, but at the same time it will usually be called without the first argument. Not sure what the convention is.